File: jobs/2026-05-27__09-48-38/auto_generate_extract_schema_py__TRJyvLM/artifacts/user/extract_task/extract.py
import os
import json
import time
from llama_cloud import LlamaCloud

def main():
    api_key = os.environ.get("LLAMA_CLOUD_API_KEY")
    run_id = os.environ.get("ZEALT_RUN_ID")
    if not api_key or not run_id:
        print("Missing environment variables.")
        return

    client = LlamaCloud(api_key=api_key)

    file_path = "/home/user/extract_task/data/invoice.pdf"
    external_file_id = f"invoice-{run_id}.pdf"

    print(f"Checking if file exists: {external_file_id}")
    files = client.files.list(external_file_id=external_file_id)
    if files.items:
        file = files.items[0]
        print(f"File already exists. ID: {file.id}")
    else:
        print(f"Uploading file: {file_path} as {external_file_id}")
        with open(file_path, "rb") as f:
            file = client.files.create(
                file=f,
                purpose="extract",
                external_file_id=external_file_id
            )
        print(f"File uploaded. ID: {file.id}")

    # Generate schema
    prompt = "Extract invoice details including invoice number, vendor name, and total amount."
    print(f"Generating schema for prompt: {prompt}")
    generated = client.extract.generate_schema(
        prompt=prompt,
        file_id=file.id
    )

    schema = generated.parameters.data_schema
    schema_path = "/home/user/extract_task/schema.json"
    with open(schema_path, "w") as f:
        json.dump(schema, f, indent=2)
    print(f"Schema saved to {schema_path}")

    # Run extraction
    print("Starting extraction job...")
    job = client.extract.create(
        file_input=file.id,
        configuration={
            "data_schema": schema,
            "extraction_target": "per_doc",
            "tier": "agentic"
        }
    )
    job_id = job.id
    print(f"Job created. ID: {job_id}")

    # Polling
    status = job.status
    while status not in ["COMPLETED", "FAILED", "CANCELLED"]:
        print(f"Current status: {status}. Waiting...")
        time.sleep(5)
        job = client.extract.get(job_id)
        status = job.status

    print(f"Terminal status reached: {status}")

    # Save result
    if status == "COMPLETED":
        result_path = "/home/user/extract_task/result.json"
        with open(result_path, "w") as f:
            json.dump(job.extract_result, f, indent=2)
        print(f"Result saved to {result_path}")
    else:
        print(f"Extraction failed with status: {status}")

    # Log file
    log_path = "/home/user/extract_task/output.log"
    with open(log_path, "w") as f:
        # Schema fields: <comma-separated property names>
        properties = schema.get("properties", {})
        fields_str = ", ".join(properties.keys())
        f.write(f"Schema fields: {fields_str}\n")
        # Job ID: <job_id>
        f.write(f"Job ID: {job_id}\n")
        # Status: COMPLETED
        f.write(f"Status: {status}\n")
    print(f"Log saved to {log_path}")

if __name__ == "__main__":
    main()

File: jobs/2026-07-02__07-42-24/llamacloud_extract_invoice_field__PX2qEsL/artifacts/user/project/extract.ts
import fs from 'fs';
import path from 'path';
import { z } from 'zod';
import LlamaCloud from '@llamaindex/llama-cloud';

// 1. Define the Zod schema for an invoice
const InvoiceSchema = z.object({
  invoice_number: z.string().describe("The unique invoice number/ID"),
  invoice_date: z.string().describe("The date of the invoice, ISO-formatted YYYY-MM-DD if possible"),
  vendor_name: z.string().describe("The name of the vendor/issuer of the invoice"),
  total_amount: z.number().describe("The grand total amount in the invoice's currency"),
  line_items: z.array(
    z.object({
      description: z.string().describe("Description of the line item"),
      quantity: z.number().describe("Quantity of the item"),
      unit_price: z.number().describe("Unit price of the item"),
      total: z.number().describe("Total cost for this line item"),
    })
  ).describe("List of individual line items in the invoice"),
});

async function main() {
  const apiKey = process.env.LLAMA_CLOUD_API_KEY;
  if (!apiKey) {
    console.error("Error: LLAMA_CLOUD_API_KEY environment variable is not set.");
    process.exit(1);
  }

  // Initialize the LlamaCloud client
  const client = new LlamaCloud({
    apiKey,
  });

  const invoicePath = '/home/user/project/invoice.pdf';
  if (!fs.existsSync(invoicePath)) {
    console.error(`Error: Input invoice file not found at ${invoicePath}`);
    process.exit(1);
  }

  console.log(`Uploading ${invoicePath} to LlamaCloud...`);
  // 2. Upload the file with purpose "extract"
  const file = await client.files.create({
    file: fs.createReadStream(invoicePath),
    purpose: 'extract',
  });

  console.log(`File uploaded successfully. File ID: ${file.id}`);

  // Convert Zod schema to JSON schema using the native v4 z.toJSONSchema method
  // @ts-ignore
  const jsonSchema = z.toJSONSchema(InvoiceSchema);
  console.log("Converted Zod schema to JSON Schema.");

  console.log("Starting Extract v2 job...");
  // 3. Start the extraction job with flattened configuration
  let job = await client.extract.create({
    file_input: file.id,
    configuration: {
      data_schema: jsonSchema as any,
      extraction_target: 'per_doc',
      tier: 'agentic',
    },
  });

  console.log(`Extract job created. Job ID: ${job.id}. Current status: ${job.status}`);

  // 4. Poll until the job is in a terminal state (COMPLETED, FAILED, or CANCELLED)
  while (job.status === 'PENDING' || job.status === 'RUNNING') {
    console.log(`Job status: ${job.status}. Polling in 5 seconds...`);
    await new Promise((resolve) => setTimeout(resolve, 5000));
    job = await client.extract.get(job.id);
  }

  console.log(`Job reached terminal state: ${job.status}`);

  if (job.status === 'COMPLETED') {
    const result = job.extract_result;
    if (!result) {
      console.error("Error: Job completed but no extract_result was returned.");
      process.exit(1);
    }

    // 5. Write the pretty-printed JSON extract_result to output.json
    const outputPath = '/home/user/project/output.json';
    fs.writeFileSync(outputPath, JSON.stringify(result, null, 2), 'utf8');
    console.log(`Successfully wrote extracted data to ${outputPath}`);

    // Safely extract the fields for the summary
    // Since result is parsed JSON, we can access its fields.
    // Let's cast it or access properties dynamically.
    const typedResult = result as any;
    const invoiceNumber = typedResult.invoice_number || 'N/A';
    const vendorName = typedResult.vendor_name || 'N/A';
    const totalAmount = typedResult.total_amount !== undefined ? typedResult.total_amount : 'N/A';

    // 6. Overwrite the one-line summary in output.log
    const logPath = '/home/user/project/output.log';
    const logLine = `Extracted Invoice: ${invoiceNumber} | Vendor: ${vendorName} | Total: ${totalAmount}\n`;
    fs.writeFileSync(logPath, logLine, 'utf8');
    console.log(`Successfully wrote summary to ${logPath}`);
  } else {
    console.error(`Extraction failed or was cancelled. Status: ${job.status}`);
    process.exit(1);
  }
}

main().catch((err) => {
  console.error("An unexpected error occurred:", err);
  process.exit(1);
});
